File: labkit/java/src/main/java/com/airamatrix/labkit/Tracer.java
package com.airamatrix.labkit;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.io.IOException;
import java.nio.file.*;
import java.util.List;
import java.util.Map;
import java.util.UUID;

/** One JSON line per agent step. Day 1 reads these; Day 4 debugs from them. */
public final class Tracer {
    private static final ObjectMapper MAPPER = new ObjectMapper();
    public final String runId;
    public final Path path;
    private final long t0 = System.currentTimeMillis();

    public Tracer(String name) {
        this.runId = name + "-" + UUID.randomUUID().toString().substring(0, 8);
        Path base = Paths.get("").toAbsolutePath();
        for (int i = 0; i < 8 && base != null; i++, base = base.getParent()) {
            if (Files.isDirectory(base.resolve("labkit"))) break;
        }
        if (base == null) base = Paths.get("").toAbsolutePath();
        Path dir = base.resolve("traces");
        try { Files.createDirectories(dir); } catch (IOException ignored) { }
        this.path = dir.resolve(runId + ".jsonl");
    }

    public void emit(String kind, Map<String, ?> fields) {
        ObjectNode record = MAPPER.createObjectNode();
        record.put("run_id", runId);
        record.put("t", (System.currentTimeMillis() - t0) / 1000.0);
        record.put("kind", kind);
        fields.forEach((k, v) -> record.set(k, MAPPER.valueToTree(v)));
        try {
            Files.writeString(path, MAPPER.writeValueAsString(record) + "\n",
                StandardOpenOption.CREATE, StandardOpenOption.APPEND);
        } catch (IOException ignored) { }
    }

    public void step(int n, String stopReason, String text, List<String> tools) {
        emit("step", Map.of("n", n, "stop_reason", String.valueOf(stopReason),
            "text", text == null ? "" : text.substring(0, Math.min(400, text.length())),
            "tools", tools));
    }

    public void tool(String name, Object args, String result, boolean ok) {
        emit("tool_result", Map.of("name", name, "args", args, "ok", ok,
            "result", result.substring(0, Math.min(400, result.length()))));
    }
}

File: labkit/python/agentic_core/tracer.py
"""Run tracing. Every step an agent takes lands here as one JSON line.

Day 1 uses this to see what the agent did. Day 4 uses the same files to debug a
failed run from its trace alone — so the observability lab is instrumented by
something participants have used since their first exercise.
"""
import json
import time
import uuid
from pathlib import Path


class Tracer:
    def __init__(self, name: str, root: str | None = None) -> None:
        self.run_id = f"{name}-{uuid.uuid4().hex[:8]}"
        base = Path(root) if root else Path(__file__).resolve().parents[3] / "traces"
        base.mkdir(parents=True, exist_ok=True)
        self.path = base / f"{self.run_id}.jsonl"
        self._t0 = time.time()

    def emit(self, kind: str, **fields) -> None:
        record = {"run_id": self.run_id, "t": round(time.time() - self._t0, 3),
                  "kind": kind, **fields}
        with self.path.open("a") as handle:
            handle.write(json.dumps(record, default=str) + "\n")

    def step(self, n: int, stop_reason: str, text: str = "", tools: list | None = None) -> None:
        self.emit("step", n=n, stop_reason=stop_reason,
                  text=(text or "")[:400], tools=tools or [])

    def tool(self, name: str, args: dict, result: str, ok: bool = True) -> None:
        self.emit("tool_result", name=name, args=args, ok=ok, result=str(result)[:400])
